tests/test_window.py

```diff
@@ -397,6 +400,85 @@ def test_doesNotTweakSpecialCharacters(self):
+        self.fakeRemoteUrl = 'https://tupac.com/i-am-alive-and-i-code-now'
+        self.testLocalImg = os.path.join(path_to_testdata, 'tupac.jpg')
+
+    @patch.object(requests, 'get')
+    def test_cacheHelper(self, requestsGet):
+        dummyImgContent = b'dummy'
+        requestsGet.return_value = MagicMock(content=dummyImgContent)
+
+	# Test that our cache helper does not impact locally hosted images.
+        localImg = "<img alt='I am alive' src='{img}'/>".format(img=self.testLocalImg)
+        self.assertEqual(
+	    self.tab.cacheRemotelyHostedImages(localImg),
+	    localImg)
+
+	# Test that our cache helper does modify remotely hosted images.
+        remoteImg = "<img alt='Come with me' src='{url}'/>".format(url=self.fakeRemoteUrl)
+        transformedText = self.tab.cacheRemotelyHostedImages(remoteImg)
+        self.assertNotEqual(
+            transformedText,
+	    remoteImg)
+
+        # Check that the new path points to one in 'tmp' (non-windows) or 'temp' (windows)
+        self.assertTrue('tmp' in transformedText or 'temp' in transformedText)
+
+        # Check hash of temp file matches hash of original content
+        matches = ReText.tab.HTML_IMG_TAG_SRC_RE.findall(transformedText)
+        self.assertEqual(
+            hashlib.md5(open(matches[0],'rb').read()).hexdigest(),
+            hashlib.md5(dummyImgContent).hexdigest())
+
+        # If we add the same image again, ensure it points to the same temporary file
+        transformedText = self.tab.cacheRemotelyHostedImages(transformedText + remoteImg)
+        matches = ReText.tab.HTML_IMG_TAG_SRC_RE.findall(transformedText)
+        self.assertEqual(
+            matches[0],
+            matches[1])
+
+        # Run the cache cleaner and make sure our temporary files are gone
+        self.tab.cleanImageCache()
+        self.assertFalse(os.path.isfile(matches[0]))
+
+    @patch('ReText.window.QMessageBox.warning', return_value='QMessageBox.Discard')
+    @patch.object(requests, 'get')
+    def test_cacheHelperWithGui(self, requestsGet, qMsg):
+        fh = open(self.testLocalImg, 'rb')
+        requestsGet.return_value = MagicMock(content=fh.read())
+
+        self.globalSettingsMock.useWebKit = True
+        self.window = ReTextWindow()
+        self.window.createNew('')
+        self.window.preview(ReText.tab.PreviewLive)
+
+        # Insert an image into the edit box, wait for the preview to be generated
+        processEventsUntilIdle()
+        time.sleep(0.5)
+        self.window.currentTab.editBox.textCursor().insertText(
+	    '![]({url})'.format(url=self.fakeRemoteUrl))
+        processEventsUntilIdle()
+        time.sleep(0.5)
+        processEventsUntilIdle()
+
+        # Check that the cache is non-empty
+        self.assertTrue(self.window.currentTab.seenImages)
+
+        # Check that we see a temporary file generated in the preview box text
+        text = self.window.currentTab.previewBox.page().mainFrame().toHtml()
```

Comment:
ReText has three preview backends, based on QTextBrowser, Qt WebKit and Qt WebEngine.

The default one is QTextBrowser, so it should be used in tests. I have fixed this now.